File: solver/Sgd.m
classdef Sgd < Optimizer
    % Stochastic gradient descent
    % sgd = Sgd(containers.Map({'learning_rate'}, {0.1}));
    
    properties(Access = protected)
        m_config = containers.Map();
        m_base_lr = 0;
    end
    
    methods(Access = public)
        % Constructor
        function obj = Sgd(config)
            obj.m_config = config;
            obj.m_base_lr = obj.m_config('learning_rate');
        end
        
        function [weights, newState] = Optimize(obj, w, dw, state)
            % Gradient descent approach (Simplest optimizer)
            w = w - (obj.m_base_lr * dw);
            weights = w;
            
            newState = [];
        end
    end
    
end

File: utils/optimization_functions/cppPrj/solver/sgd.cpp
#include "sgd.h"
#include <map>
template <typename T>
SGD<T>::SGD(const map<string, float> &config):BaseOptimizer<T>(config){
    // Ugly detail about templated classes (https://isocpp.org/wiki/faq/templates#nondependent-name-lookup-members)

    // try to find key "learning_rate" returns iterator if found
    auto search = BaseOptimizer<T>::m_config.find("learning_rate");

    // if we found "learning_rate" dereference the iterator to get the associated value
    if (search != BaseOptimizer<T>::m_config.end()){
        m_base_lr = search->second;
    }
    else{
        throw("No learning rate set");
    }

    //m_base_lr = (*BaseOptimizer<T>::m_config.find("earning_rate")).second;
    //m_base_lr = BaseOptimizer<T>::m_config.find("learning_rate")->second;

}

template<typename T>
Tensor<T> SGD<T>::Optimize(const Tensor<T> &params, const Tensor<T> &grad_params, const OptimizerState<T> &state){

    // Gradient descent approach (Simpliest optimizer, no use of states)
    auto weights = params - (grad_params*m_base_lr);

    return weights;
}

/*
 * Explicit declare template versions to avoid linker error. (This is needed if we use templates on .cpp files)
*/
template class SGD<float>;
template class SGD<double>;
template class SGD<int>;

